src/backend/distributed/planner/modify_planner.c

```diff
@@ -641,3 +644,315 @@ ExtractPartitionValue(Query *query, Var *partitionColumn)
+	 */
+	if (query->hasSubLinks == true || query->cteList != NIL)
+	{
+		return false;
+	}
+
+
+	/* only hash partitioned tables are supported */
+	distributedTableId = ExtractFirstDistributedTableId(query);
+	partitionColumn = PartitionColumn(distributedTableId, rangeTableId);
+	partitionMethod = PartitionMethod(distributedTableId);
+
+	if (partitionMethod != DISTRIBUTE_BY_HASH)
+	{
+		return false;
+	}
+
+	/* extract range table entries */
+	ExtractRangeTableEntryWalker((Node *) query, &rangeTableList);
+
+	/* query can have only one range table of type RTE_RELATION */
+	if (list_length(rangeTableList) != 1)
+	{
+		return false;
+	}
+	else
+	{
+		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) linitial(rangeTableList);
+		if (rangeTableEntry->rtekind != RTE_RELATION)
+		{
+			return false;
+		}
+	}
+
+	if (joinTree == NULL)
+	{
+		return false;
+	}
+
+	if ((joinTree != NULL) && (joinTree->quals != NULL))
+	{
+		quals = joinTree->quals;
+		if (IsA(quals, List))
+		{
+			quals = (Node *) make_ands_explicit((List *) quals);
+		}
+	}
+
+	partitionColumnMatchExpression =
+		ContainsPartitionColumnMatchExpression(quals, partitionColumn);
+
+	if (!partitionColumnMatchExpression)
+	{
+		return false;
+	}
+
+	/* make sure partition column is used only once in the query */
+	varClauseList = pull_var_clause_default(quals);
+	foreach(varClauseCell, varClauseList)
+	{
+		Var *column = (Var *) lfirst(varClauseCell);
+		if (equal(column, partitionColumn))
+		{
+			partitionColumnReferenceCount++;
+		}
+	}
+
+	if (partitionColumnReferenceCount != 1)
+	{
+		return false;
+	}
+
+	shardCount = ShardIntervalCount(distributedTableId);
+
+	if (shardCount == 0)
+	{
+		return false;
+	}
+
+	return true;
+}
+
+
+/* MultiRouterPlanCreate creates a physical plan for given router plannable query. */
+MultiPlan *
+MultiRouterPlanCreate(Query *query)
+{
+	Task *task = NULL;
+	Job *job = NULL;
+	MultiPlan *multiPlan = NULL;
+
+	task = SingleShardSelectTask(query);
+
+	job = SingleShardSelectJob(query,task);
+
+	multiPlan = CitusMakeNode(MultiPlan);
+	multiPlan->workerJob = job;
+	multiPlan->masterQuery = NULL;
+	multiPlan->masterTableName = NULL;
+
+	return multiPlan;
+}
+
+
+/* SingleShardSelectTask builds a Task to represent a single shard select query */
+static Task *
+SingleShardSelectTask(Query *query)
+{
+	Task *task = NULL;
+	ShardInterval *shardInterval = SingleShardSelectShardInterval(query);
+	StringInfo queryString = makeStringInfo();
+	uint64 shardId = INVALID_SHARD_ID;
+	bool upsertQuery = false;
+	CmdType commandType = query->commandType;
+	FromExpr *joinTree = NULL;
+
+	Assert(shardInterval != NULL);
+	Assert(commandType == CMD_SELECT);
+
+	shardId = shardInterval->shardId;
+
+	/*
+	 * Convert the qualifiers to an explicitly and'd clause, which is needed
+	 * before we deparse the query.
+	 */
+	joinTree = query->jointree;
+	if ((joinTree != NULL) && (joinTree->quals != NULL))
+	{
+		Node *whereClause = joinTree->quals;
+		if (IsA(whereClause, List))
```

Comment:
Do we need this `if`? Can't we do something like Citus has [here](https://github.com/citusdata/citus/blob/ac05e561ed5f4e2ccc0562d5677a5253bceca88b/src/backend/distributed/planner/multi_physical_planner.c#L2097)?